Error when manually including vite/modulepreload-polyfill

### Describe the bug

When manually including `vite/modulepreload-polyfill` as per [backend integration](https://vitejs.dev/guide/backend-integration.html) an error is thrown in the browser when running `vite serve`. 

```
Uncaught ReferenceError: __VITE_IS_MODERN__ is not defined
    <anonymous> http://localhost:3000/@id/vite/modulepreload-polyfill:43
```

I'm guessing because `__VITE_IS_MODERN__` is replaced during build and during "normal" Vite usage the polyfill is only added to the output during build, but with a backend integration the polyfill is present during dev and build.

### Reproduction

Issue can be replicated outside of a backend integration by:

1. Importing `vite/modulepreload-polyfill` into an entry file
2. Running `vite serve`, visiting the dev server and opening the browser dev tools

This yields the same result as I'm seeing with my backend setup. 

https://github.com/rynpsc/vite-modulepreload-polyfill

### System Info

```shell
System:
    OS: macOS 11.5.2
    CPU: (8) x64 Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
    Memory: 681.98 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
Binaries:
    Node: 12.18.3 - ~/.nvm/versions/node/v12.18.3/bin/node
    npm: 7.5.2 - ~/.nvm/versions/node/v12.18.3/bin/npm
Browsers:
    Edge: 92.0.902.84
    Firefox: 91.0.1
    Safari: 14.1.2
```


### Used Package Manager

npm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

Unfortunately in TypeScript, the workaround throws the error: `Cannot find module 'vite/modulepreload-polyfill' or its corresponding type declarations.`

My workaround was to copy the polyfill function at https://github.com/vitejs/vite/blob/cb75dbd4a2b1c56e2a7aeaf1625818a4d1865f00/packages/vite/src/node/plugins/modulePreloadPolyfill.ts and run it in its own file.

--- a/packages/playground/backend-integration/__tests__/backend-integration.spec.ts
+++ b/packages/playground/backend-integration/__tests__/backend-integration.spec.ts
@@ -32,6 +32,12 @@ if (isBuild) {
     expect(htmlEntry.assets.length).toEqual(1)
   })
 } else {
+  test('No ReferenceError', async () => {
+    browserErrors.forEach((error) => {
+      expect(error.name).not.toBe('ReferenceError')
+    })
+  })
+
   describe('CSS HMR', () => {
     test('preserve the base in CSS HMR', async () => {
       await untilUpdated(() => getColor('body'), 'black') // sanity check

--- a/packages/playground/backend-integration/frontend/entrypoints/main.ts
+++ b/packages/playground/backend-integration/frontend/entrypoints/main.ts
@@ -1,3 +1,5 @@
+import 'vite/modulepreload-polyfill'
+
 export const colorClass = 'text-black'
 
 export function colorHeading() {

--- a/packages/vite/src/node/plugins/modulePreloadPolyfill.ts
+++ b/packages/vite/src/node/plugins/modulePreloadPolyfill.ts
@@ -5,7 +5,8 @@ import { isModernFlag } from './importAnalysisBuild'
 export const modulePreloadPolyfillId = 'vite/modulepreload-polyfill'
 
 export function modulePreloadPolyfillPlugin(config: ResolvedConfig): Plugin {
-  const skip = config.build.ssr
+  // `isModernFlag` is only available during build since it is resolved by `vite:build-import-analysis`
+  const skip = config.command !== 'build' || config.build.ssr
   let polyfillString: string | undefined
 
   return {

--- a/scripts/jestPerTestSetup.ts
+++ b/scripts/jestPerTestSetup.ts
@@ -24,6 +24,7 @@ declare global {
   const page: Page | undefined
 
   const browserLogs: string[]
+  const browserErrors: Error[]
   const serverLogs: string[]
   const viteTestUrl: string | undefined
   const watcher: RollupWatcher | undefined
@@ -34,6 +35,7 @@ declare const global: {
   page?: Page
 
   browserLogs: string[]
+  browserErrors: Error[]
   serverLogs: string[]
   viteTestUrl?: string
   watcher?: RollupWatcher
@@ -56,6 +58,11 @@ const onConsole = (msg: ConsoleMessage) => {
   logs.push(msg.text())
 }
 
+const errors: Error[] = (global.browserErrors = [])
+const onPageError = (error: Error) => {
+  errors.push(error)
+}
+
 beforeAll(async () => {
   const page = global.page
   if (!page) {
@@ -63,6 +70,7 @@ beforeAll(async () => {
   }
   try {
     page.on('console', onConsole)
+    page.on('pageerror', onPageError)
 
     const testPath = expect.getState().testPath
     const testName = slash(testPath).match(/playground\/([\w-]+)\//)?.[1]

